### DISCUSSION TITLE
How to test for ImportError or ModuleNotFoundError being handled correctly?

### DISCUSSION THREAD

**Discussion by gothicVI** [pytest-dev/pytest#12864] [Q&A] [answered]
I have a module that has in its `__init__.py` the following code:
```
import logging
import sys

logger: logging.Logger = logging.getLogger(__name__)

try:
    from . import (
        a,
        b,
        c,
    )
except ModuleNotFoundError:
    logger.critical("Can't import from .!")
    sys.exit(1)

__all__: list[str] = ["ea", "b", "c"]
```
The structure of the module is as follows:
```
.
├── src
│   └── mymodule
│       ├── a.py
│       ├── b.py
│       ├── c.py
│       └── __init__.py
└── tests
```
How can I test that a `ImportError` or `ModuleNotFoundError` would write the correct log message and exit with exit code `1`?

**Comment by RonnyPfannschmidt** [comment]
That particular kind of import hack is  best tested by using a subprocess 

Personally id strongly recomend to avoid that pattern

With more context i may be able to recommend a more testable pattern

**Reply by gothicVI** [reply]
I've updated the structure of the project above.
Can you elaborate on why that pattern is bad? I've seen it in many projects.
Sure, strictly speaking it's not necessary as everything gets packaged together but I was under the expression that importing local modules should be wrapped in a `try... except...`.

**Reply by RonnyPfannschmidt** [reply, marked as answer]
Putting imports into except is only done in case of externals being optional

Its never the default
